File: Leetcode算法题/linked_list/add_two_numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        
        carry = 0
        num = ""
        current_1 = l1
        current_2 = l2
        
        while current_1 is not None or current_2 is not None or carry != 0:
            sum = ((current_1.val if current_1 is not None else 0) + (current_2.val if current_2 is not None else 0) + carry)
            carry = 0
            num += str(sum%10)
            carry += 0 if sum < 10 else 1
            current_1 = current_1.next if current_1 is not None else None
            current_2 = current_2.next if current_2 is not None else None
            
        l3 = None
            
        for c in num[::-1]:
            l3 = ListNode(c, l3)
        
        return l3

File: Leetcode算法题/linked_list/add_two_numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
    
    string reverseString(string s) {
        string temp = "";

        for (int i = s.length() - 1; i >= 0; i--) {
            temp += s[i];
        }

        return temp;
    }
    
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* l3 = NULL;
        int carry = 0;
        string num = "";
        ListNode* current_1 = l1;
        ListNode* current_2 = l2;
        
        while (current_1 != NULL || current_2 != NULL || carry > 0) {
            int sum = (current_1 != NULL ? current_1->val : 0) + (current_2 != NULL ? current_2->val : 0) + carry;
            carry = 0;
            num += to_string(sum % 10);
            carry += sum < 10 ? 0 : 1;
            current_1 = current_1 != NULL ? current_1->next : NULL;
            current_2 = current_2 != NULL ? current_2->next : NULL;
        }
        
        for (char& c : reverseString(num)) {
            l3 = new ListNode(c-48, l3);
        }
        
        return l3;
    }
};
